qcodes/instrument/parameter.py

```diff
@@ -94,83 +94,89 @@ class _BaseParameter(Metadatable, DeferredOperations):
             ``update=True``, for example if it takes too long to update.
             Default True.
 
+        max_val_age (Optional[int]): The max time (in seconds) to trust a
+            saved value obtained from get_latest(). If this parameter has not
+            been set or measured more recently than this, perform an
+            additional measurement.
+
         metadata (Optional[dict]): extra information to include with the
             JSON snapshot of the parameter
     """
     def __init__(self, name, instrument, snapshot_get, metadata,
-                 snapshot_value=True):
+                 step=None, scale=None, inter_delay=0, post_delay=0,
+                 val_mapping=None, get_parser=None, set_parser=None,
+                 snapshot_value=True, max_val_age=None, vals=None):
         super().__init__(metadata)
-        self._snapshot_get = snapshot_get
         self.name = str(name)
         self._instrument = instrument
+        self._snapshot_get = snapshot_get
         self._snapshot_value = snapshot_value
 
-        self.has_get = hasattr(self, 'get')
-        self.has_set = hasattr(self, 'set')
+        if not isinstance(vals, (Validator, type(None))):
+            raise TypeError('vals must be None or a Validator')
+        self.vals = vals
+
+        self.step = step
+        self.scale = scale
+        self.raw_value = None
+        self.inter_delay = inter_delay
+        self.post_delay = post_delay
 
-        if not (self.has_get or self.has_set):
-            raise AttributeError('A parameter must have either a get or a '
-                                 'set method, or both.')
+        # TODO (nulinspiratie) handle int vs string conversion in val_mapping
+        self.val_mapping = val_mapping
+        if val_mapping is None:
+            self.inverse_val_mapping = None
+        else:
+            self.inverse_val_mapping = {v: k for k, v in val_mapping.items()}
+
+        self.get_parser = get_parser
+        self.set_parser = set_parser
 
         # record of latest value and when it was set or measured
         # what exactly this means is different for different subclasses
         # but they all use the same attributes so snapshot is consistent.
-        self._latest_value = None
-        self._latest_ts = None
-        self.get_latest = GetLatest(self)
+        self._latest = {'value': None, 'ts': None}
+        self.get_latest = GetLatest(self, max_val_age=max_val_age)
+
+        if hasattr(self, 'get'):
```

Comment:
The other way I can think of is to use it as a wrapper (e.g. place `@wrap_get` above `def get`). However, the issue here is that anyone who creates a custom parameter must remember to place it before each get and set, otherwise all the attributes that are in the wrapper won't have any effect. 

In case someone explicitly does not want the wrapping functionality, it can always be overridden afterwards by redefining its `get/set` method. (e.g. `self.get = CustomParameter.get`). I thought this was the better solution, as the wrapping is enabled by default. 

Potentially a metaclass could also handle this, but I thought it would add complexity and be unnecessary. But if anyone has another way please let me know